Update Snakemake template & docs with 'run_<challenge>' flags

SamBryce-Smith opened the pull request:
Partially fixes #189

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated corresponding READMEs (if applicable)
- [ ] My code follows the templates/style guidelines of the repository
- [ ] In- and output formats comply with APAeval specifications
- [ ] No parameters or file names are hardcoded
- [ ] Results, logs or other output is not commited to the repository
- [ ] I have tested my code

First attempt at updating the Snakemake template to have flags in config to switch on/off output for each challenge.

To double check I had the syntax right I made a few modifications so that the template can successfully dry run. If you're sitting in the `snakemake` directory (`docs/templates/snakemake`), you can dry run with the following steps:
```
# 1. Activate the APAeval execution conda environment
## If you haven't installed the environment yet, can do so with conda env create -f ../../../apaeval_env.yaml
$ conda activate apaeval_execution_workflows

#2. Perform dry run
$ snakemake -n -p --configfile config/config.\[METHOD\].yaml
```

I've also made updates to the README accordingly, trying to mirror the Nextflow updates as much as possible (#196 - it would probably be easier if these were on the same PR...). I tried to remove any references to conda environments as we don't want our workflows to use these anymore.

Feedback would be much appreciated :)

dominikburri reviewed (commented):
Hi Sam, thanks a lot for the work! 

The dry run and rule graph worked like a charm! After activating the conda env, I executed `dryrun.sh` and `rulegraph.sh` and both executed successfully. 
Is the idea also to test `run_local.sh`? 

I think the next step is to test your construct on an ewf that is already merged. What do you think?

SamBryce-Smith commented on docs/templates/snakemake/README.md:
```diff
@@ -37,12 +41,14 @@ Read this section, but do NOT include it your final README.
 
 {Describe parameters needed by your METHOD.}
 
+{Describe the parameters for running different modes `run_identification`, `run_quantification`, `run_differential`}
```
Good point I will definitely add this clarification

SamBryce-Smith replied:
Added in 80b4222 (I think I added it in the input section as that is where sample tables are first mentioned)

SamBryce-Smith commented on docs/templates/snakemake/workflow/Snakefile:
```diff
@@ -7,6 +7,9 @@ import pandas as pd
 samples = pd.read_csv(os.path.abspath(
     config["samples"])).set_index("sample", drop=False)
 
+assert isinstance(config["run_identification"], bool), f"'run_identification' must be True/False boolean, {config['run_identification']} (type {type(config['run_identification'])}) was provided"
+assert isinstance(config["run_quantification"], bool), f"'run_quantification' must be True/False boolean, {config['run_quantification']} (type {type(config['run_quantification'])}) was provided"
+assert isinstance(config["run_differential"], bool), f"'run_differential' must be True/False boolean, {config['run_differential']} (type {type(config['run_differential'])}) was provided"
```
> Maybe add an extra step that reports missing flags to make sure that it is the desired behaviour. Something like: "INFO: The flag run_differential is missing. If this method does not support it, you can disregard this message."

Yes I think this is a great idea if we go with the remove the flag option.

> Maybe I got the comment wrong and I should not remove the whole line in the first place. In that case your assert statements would correctly catch that.


No you are absolutely correct, I think elsewhere in the README I said to remove the flag if the tool is not compatible with that challenge, so following from that the assert would fail if it's not removed. I agree I'm also not sure what the best option is (I've switched preferences as I've written this reply...), I think both are sensible and we'll have to decide and stick with one.

Some thoughts in favour of removing the flag altogether:

- Don't include a parameter that cannot be changed. (_maybe_ it could cause some confusion/room for accidental changes if retained in the config (though with an assert statement the inconvenience would be negligible))
- Less code in the Snakemake template (i.e. don't need to check a parameter that has no effect on the workflow)
- Don't 'suggest' a tool can perform a task is does not do (although enforce False with info message would cover this anyway)

Thoughts in favour of enforce False:
- If changes to challenges that mean tool is now compatible (e.g. a 'relative quantification' benchmark), the flag is still there (just need to remove enforce False assert statement and get Snakemake to track the output file)


I'm wondering whether we should make a distinction between the following:
- tool can perform a task **but does not fit our challenge specifications**
    - Enforce the flag to False (assert statement to check, perhaps an info message too?)
-  tool does not perform a given task at all
    - Remove the flag entirely (with the missing key info message you suggested)


My main worry with above suggestion is this template (and corresponding documentation) becomes convoluted and difficult to follow for new users (though at this stage of the project there won't be many of those...)

I feel like any option will be equally valid so I'm open to input/a decision. I think the final option is most comprehensive but would add a layer of complexity that we may not want at this stage...

dominikburri replied:
Discussed in a previous meeting, see comment for outcome below.

SamBryce-Smith commented on docs/templates/snakemake/config/config.[METHOD].yaml:
```diff
@@ -11,6 +11,27 @@ method: "METHOD" # same as the name of the directory
+
+run_differential: True
+
+identification_output_suffix: ".identification.bed"
+quantification_output_suffix: ".quantification.bed"
+differential_output_file: "differential_challenge.tsv"
```
Hmm, There wasn't much thought in the specific values other than to get something dry-runnable. I chose output suffixes for ID & quant mainly because the Ewfs and run specifications I've seen so far are per-sample, so just wanted a way to differentiate between the different challenge files. I guess this is based on the assumption all challenge output files will go to the same output directory, but that may not actually reflect our documentation/how we plan to run each challenge. 

I now realise I should be very careful here as I don't want to introduce conflicts between the documentation and the template. Judging from [output filenames specifications](https://github.com/iRNA-COSI/APAeval/tree/main/execution_workflows#filenames), we want our output filenames to be `CHALLENGECODE_PARTICIPANT_OUTCODE.ext`, where:

- CHALLENGECODE - in my head analogous to a sample ID, in which case defined in `samples.csv`?
- PARTICIPANT - name of tool/workflow (need to add this to the config)
- OUTCODE - [01 for ID, 02 for quantification & 03 for differential](https://github.com/iRNA-COSI/APAeval/blob/main/execution_workflows/execution_output_specification.md#outputs)
 
so I need to update the template & `config.yaml` to reflect this.

With that in mind I should also update the `*_output_suffix` parameters to be:
- "{challenge code}.bed" for ID & quant, where challenge codes are 01 & 02 respectively
- "03.tsv" for differential challenge (replacing `differential_output_file` with `differential_output_suffix`)

I could also have the challenge codes and extensions as separate parameters if we want more customisation, but ultimately they'd be joined together anyway. In my head 'challenge code' is analogous to a sample ID, in which case it would be defined in `samples.csv`, but that may not be the case for all of our challenges.

dominikburri replied:
Thanks for the explanations :) 
I think the current state is good, because it is flexible and easy to adjust. So I don't think we need to do something here, just to keep in mind how we want to save the runs in the end.

SamBryce-Smith commented on docs/templates/snakemake/README.md:
```diff
@@ -8,21 +8,25 @@ Read this section, but do NOT include it your final README.
+>     * Each rule has it's own [Docker container](https://www.docker.com/resources/what-container). All containers should be pushed to the [APAeval team Dockerhub](https://hub.docker.com/u/apaeval)
 > * There are some shell scripts which can be used to start a snakemake run. Adjust the name of the `config file` in these scripts:
 >     * `dryrun.sh`
 >     * `rulegraph.sh` (Also adjust name of output `.png`)
```
Added in 80b4222

SamBryce-Smith commented on docs/templates/snakemake/config/config.[METHOD].yaml:
```diff
@@ -11,6 +11,27 @@ method: "METHOD" # same as the name of the directory
+# If combinations of flags are incompatible with one another (e.g. run_differential requires a different workflow to run_quantification), this should be noted here and in the README
+
+
+# Whether to run steps to produce output files for identification challenge
```
I included for the other two flags completely out of personal preference 80b4222

SamBryce-Smith commented:
Have finally got around to addressing your first round of suggestions @dominikburri . I've also updated the main Snakefile and README to reflect our decision from previous meeting (always keep all 3 flags, enforce to False if incompatible). I also added quick info messages on whether challenge flags have been set to T/F & whether output will be generated (I think both @dominikburri  & @mrgazzara suggested this). `dryrun.sh` & `rulegraph.sh` both still work with the template as is.

Comments on the above are welcome as always! I can test this on DaPars2 in due cource (as already merged), but I'm currently prioritising working on the Roar workflow to get that ready to review).

dominikburri reviewed (approved):
Thank you for the work!

faricazjj reviewed (approved):
looks good to me! :D thanks a lot for this @SamBryce-Smith!!

SamBryce-Smith commented:
Thanks @dominikburri  & @faricazjj , now merged to main!
